presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDecimalSelectiveStreamReader.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.reader;
+
+import com.facebook.presto.memory.context.LocalMemoryContext;
+import com.facebook.presto.orc.StreamDescriptor;
+import com.facebook.presto.orc.TupleDomainFilter;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.Int128ArrayBlock;
+import com.facebook.presto.spi.type.UnscaledDecimal128Arithmetic;
+import io.airlift.slice.Slice;
+import io.airlift.slice.UnsafeSlice;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import static com.facebook.presto.spi.type.UnscaledDecimal128Arithmetic.rescale;
+
+public class LongDecimalSelectiveStreamReader
+        extends AbstractDecimalSelectiveStreamReader
+{
+    public LongDecimalSelectiveStreamReader(
+            StreamDescriptor streamDescriptor,
+            Optional<TupleDomainFilter> filter,
+            boolean outputRequired,
+            LocalMemoryContext systemMemoryContext)
+    {
+        super(streamDescriptor, filter, outputRequired, systemMemoryContext);
+    }
+
+    @Override
+    int readNoFilter(int[] positions, int positionCount)
+            throws IOException
+    {
+        Slice decimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
+        Slice rescaledDecimal = UnscaledDecimal128Arithmetic.unscaledDecimal();
+
+        int streamPosition = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int position = positions[i];
+            if (position > streamPosition) {
+                skip(position - streamPosition);
+                streamPosition = position;
+            }
+
+            if (presentStream != null && !presentStream.nextBit()) {
+                if ((nonDeterministicFilter && filter.testNull()) || nullsAllowed) {
```

Comment:
- since this is a `readNoFilter` method, filter is always `null`, nullsAllowed is always `true` and there is no need to compute outputPositionCount; a simplified version could look like this:

```
        int streamPosition = 0;
        for (int i = 0; i < positionCount; i++) {
            int position = positions[i];
            if (position > streamPosition) {
                skip(position - streamPosition);
                streamPosition = position;
            }

            if (presentStream != null && !presentStream.nextBit()) {
                nulls[i] = true;
            }
            else {
                int scale = (int) scaleStream.next();
                dataStream.nextLongDecimal(decimal);
                rescale(decimal, decimalType.getScale() - scale, rescaledDecimal);
                values[2 * i] = UnsafeSlice.getLongUnchecked(rescaledDecimal, 0);
                values[2 * i + 1] = UnsafeSlice.getLongUnchecked(rescaledDecimal, Long.BYTES);
                if (presentStream != null) {
                    nulls[i] = false;
                }
            }

            streamPosition++;
        }
        outputPositionCount = positionCount;
        return streamPosition;
```